cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataScanner.java

```diff
@@ -0,0 +1,236 @@
+import co.cask.cdap.logging.meta.LoggingStoreTableUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import org.apache.tephra.TransactionFailureException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class to scan and also delete meta data
+ */
+public class FileMetadataScanner {
+  private static final Logger LOG = LoggerFactory.getLogger(FileMetadataScanner.class);
+  private static final byte[] OLD_ROW_KEY_PREFIX = LoggingStoreTableUtil.OLD_FILE_META_ROW_KEY_PREFIX;
+  private static final byte[] OLD_ROW_KEY_PREFIX_END = Bytes.stopKeyForPrefix(OLD_ROW_KEY_PREFIX);
+
+  private static final byte[] NEW_ROW_KEY_PREFIX = LoggingStoreTableUtil.NEW_FILE_META_ROW_KEY_PREFIX;
+  private static final byte[] NEW_ROW_KEY_PREFIX_END = Bytes.stopKeyForPrefix(NEW_ROW_KEY_PREFIX);
+
+  // cut-off time discount from actual transaction timeout
+  private static final int CUTOFF_DISCOUNT = 5;
+  private final Transactional transactional;
+  private final DatasetManager datasetManager;
+
+  public FileMetadataScanner(DatasetManager datasetManager, Transactional transactional) {
+    this.transactional = transactional;
+    this.datasetManager = datasetManager;
+  }
+
+  /**
+   * scans for meta data in old format which has expired the log retention.
+   */
+  @VisibleForTesting
+  List<byte[]> scanAndDeleteOldMetaData(int transactionTimeout, final int cutoffTransactionTime) {
+    final List<byte[]> deletedEntries = new ArrayList<>();
+    try {
+      transactional.execute(transactionTimeout, new TxRunnable() {
+        public void run(DatasetContext context) throws Exception {
+          Stopwatch stopwatch = new Stopwatch();
+          stopwatch.start();
+          Table table = LoggingStoreTableUtil.getMetadataTable(context, datasetManager);
+          // create range with tillTime as endColumn
+          Scan scan = new Scan(OLD_ROW_KEY_PREFIX, OLD_ROW_KEY_PREFIX_END, null);
+          try (Scanner scanner = table.scan(scan)) {
+            Row row;
+            while ((row = scanner.next()) != null) {
+              if (stopwatch.elapsedTime(TimeUnit.SECONDS) > cutoffTransactionTime) {
+                break;
+              }
+              byte[] rowKey = row.getRow();
+              // delete all columns for this row
+              table.delete(rowKey);
+              deletedEntries.add(rowKey);
+            }
+          }
+        }
+      });
+    } catch (TransactionFailureException e) {
+      LOG.warn("Got Exception while deleting old metadata", e);
+      // not required as we dont delete files in old format, but still safe to clear.
+      deletedEntries.clear();
+    }
+    LOG.info("Deleted {} entries from the meta table for the old log format", deletedEntries.size());
+    return deletedEntries;
```

Comment:
Does anyone using this list? If not, no need to return it (also returning the row key is not good anyway as it requires the caller, if decided to use it, to understand and decode the row key)